File: 1-4-palindrome-permutation.py
#  1.4 Palindrome Permutation

#  Given a string, write a function to check if it is a permutation of a palindrome. 
#  A palindrome is a word or phrase that is the same forwards and backwords. 
#  A permutation is a rearrangement of letters.
#  A palindrome does not need to be limited to just dictionary words.

def is_palindrome_permutation(string):

    # Let's assume that palindromes must be at least three characters long.
    length = len(string)
    if length <= 2:
        return False

    # Count the frequency of characters in a string.
    seen = {}
    for c in string:
        if c in seen:
            seen[c] += 1
        else:
            seen[c] = 1

    # If a palindrome has an even length, all characters must occur an even number of times.
    if length % 2 == 0:
        for char in seen:
            if seen[char] % 2 != 0:
                return False
        return True

    # If a palindrome is an odd length, one character must occur an odd number of times, 
    # while all other characters must occur an even number of times.
    if length % 2 == 1:
        odd_frequency_seen = False
        for char in seen:
            if seen[char] % 2 != 0:
                if odd_frequency_seen:
                    return False
                else:
                    odd_frequency_seen = True
        if not odd_frequency_seen:
            return False
        else:
            return True

print(is_palindrome_permutation('deed'))  # True
print(is_palindrome_permutation('civic'))  # True
print(is_palindrome_permutation('Max Wiederholt'))  # False

File: 1-4-palindrome-permutation.js
// 1.4 Palindrome Permutation

// Given a string, write a function to check if it is a permutation of a palindrome. 
// A palindrome is a word or phrase that is the same forwards and backwards. (E.e., "deed", "civic", "madam", "radar", "level", "Stanley Yelnats")
// A permutation is a rearrangement of letters. The palindrome does not need to be limited to just dictionary words.
// In this implementation, we assume a user has provided a string with no spaces.

function is_palindrome_permutation(s)
{
    // Create a dictionary where the keys are unique characters in the given strings 
    // and the values are the number of times that character is seen in the string.
    let character_counts = {};
    for(let i = 0, length = s.length; i < length; i++)
    {
        if(character_counts[s[i]])
        {
            character_counts[s[i]] += 1;
        }
        else
        {
            character_counts[s[i]] = 1;
        }
    }

    // If the given string has an odd length, exactly one character must have an odd character count.
    if (s.length % 2 == 1)
    {
        let odd_count_seen = false;
        for (let count in character_counts)
        {
            if(character_counts[count] % 2 == 1)
            {
                if (odd_count_seen)
                {
                    return false;
                }
                else
                {
                    odd_count_seen = true;
                }
            }
        }
        if (odd_count_seen)
        {
            return true;
        }
        else
        {
            return false;
        }        
    }

    // If the given string has an even length, all characters must have an even character count.
    else
    {
        for (let count in character_counts)
        {
            if(character_counts[count] % 2 == 1)
            {
                return false;
            }
        }        
    }

    return true;
}

console.log(
    is_palindrome_permutation("civic")
);  // true

console.log(
    is_palindrome_permutation("deed")
);  // true

console.log(
    is_palindrome_permutation("Max Wiederholt")
);  // false